# 코루틴 스코프

코루틴 스코프: `스코프 범위 내에서 생성된 코루틴에게 실행 환경을 제공하고, 이 코루틴의 동작을 관리`한다.

코루틴의 실행 환경을 저장하고 관리하는 CoroutineContext 를 가지고 있다.

코루틴 스코프에 있는 CoroutineContext 에 만약 IO Dispatcher 가 존재한다면 해당 코루틴 스코프 내에서 만들어진 코루틴들은
모두 IO Dispatcher 에 의해 IO 스레드에서 동작하게 될 것이다.

그리고, CoroutineScope 에 있는 Job 이 있으면 모든 코루틴들은 이 Job 을 root Job 으로 삼고 있을 것이다.

확인해보자

```kotlin
import kotlin.coroutines.coroutineContext

suspend fun main() {
    val scope = object : CoroutineScope {
        override val coroutineContext: CoroutineContext = Dispatchers.Default + Job()
    }
    
    // launch 로 인해 생성되는 코루틴에게 coroutineContext(실행 환경) 을 제공한다.
    
    scope.launch {
        println(coroutineContext)
    }
}
```

# CoroutineScope() 팩토리 함수를 통해 스코프를 만들자

기본적으로 `CoroutineScope()` 팩토리 함수에는 params로 받는 context 에 Job 이 없으면 Job() 을 통해 root Job을 만들고 있다.

# 커스텀 CoroutineScope 를 만들지 말아라 !

만약, 커스텀 CoroutineScope 를 만들면 어떻게 될까?

```kotlin
suspend fun main() {
    val scope = object : CoroutineScope {
        override val coroutineContext: CoroutineContext = Dispatchers.Default
    }

    scope.launch {
        println(coroutineContext[Job]) // StandaloneCoroutine{Active}@4fba6dc8
        println(coroutineContext[Job]?.parent) // null
    }
    delay(100)
}
```
CoroutineScope 에 rootJob 이 존재하지 않게된다.
그럼 CoroutineScope 의 책임 중 하나인 코루틴의 동작을 관리할 수 없게 된다.

```kotlin
suspend fun main() {
    val scope = object : CoroutineScope {
        override val coroutineContext: CoroutineContext = Dispatchers.Default
    }

    scope.cancel()
}
```
만약, scope 의 cancel() 을 호출해보자

바로 에러 발생..
```
Exception in thread "main" java.lang.IllegalStateException: Scope cannot be cancelled because it does not have a job
```


CoroutineScope 는 스코프 내에서 생성된 코루틴들에게 실행 환경을 제공하고 이들의 실행 범위를 제어하는 역할
그래서 coroutineContext를 홀딩하고 있다
1) 코루틴들에게 실행 환경을 제공
    - Dispatcher - corotineContext 내부의 CoroutineDispatcher 가 launch 코루틴이 어느 스레드에서 돌아가도록할지 배분한다.
    - 부모 Job 제공
    -
2) 실행 범위를 제어하는 역할